### PR TITLE
MDEV-28153: Debian autobake- use absolute dependencies

### PR DESCRIPTION
**Pull request by grooverdan**
- [x] *The Jira issue number for this PR is: MDEV-28153*

## Description

With autobake-deb.sh no doing explicit dependencies, we need to run this
explicitly as a prep step in salsa-ci so the dependencies can be pulled
before the real compilation begins.

Resolves 10.5 part of salsa in #2039.

## How can this PR be tested?

Currently running on https://salsa.debian.org/grooverdan/mariadb-server/-/pipelines/361865)

(10.8 test timed out)

## Basing the PR against the correct MariaDB version
- [X] *This is a bug fix and the PR is based against the earliest branch in which the bug can be reproduced*

## Backward compatibility
TODO: fill details here, if applicable, or remove the section

### CODE REVIEW

**Review by ottok — CHANGES_REQUESTED** [review]
I don't understand the commit message / documentation. Why are these changes needed? What is broken - what do these changes fix?

What is _early bit of autobake-deb.sh_ that needs lsb-release? What is this SALSA_PREP thing? Why is columnstore stanza refactored?

**Comment by ottok** [comment]
Changed
```
## Basing the PR against the correct MariaDB version
- [ ] *This is a new feature and the PR is based against the latest MariaDB development branch*
- [ X ] *This is a bug fix and the PR is based against the earliest branch in which the bug can be reproduced*
```
to
```
## Basing the PR against the correct MariaDB version
- [X] *This is a bug fix and the PR is based against the earliest branch in which the bug can be reproduced*
```

**Comment by grooverdan** [comment]
> I don't understand the commit message / documentation. Why are these changes needed?

Since MDEV-28153, debian dependencies are now a property of the script. Without adjusting control before `apt-get build-dep`, an incorrect, possibly impossible, set of dependencies are generated.
 
> What is broken - what do these changes fix?

By running the prepare only of autobake-deb.sh, the control file has its build dependencies updated.  Running `apt-get build-dep` on any system will now pull the correct acheivable dependencies.

> What is _early bit of autobake-deb.sh_ that needs lsb-release?

`CODENAME="$(lsb_release -sc)"`

> What is this SALSA_PREP thing?

A flag to run the prepare stage only, so that `control`  (and `rules`) files only are updated.

> Why is columnstore stanza refactored?

It need to be id-potent so that a second run to trigger the build doesn't get two sets of included columnstore files, especially in `debian/control`

**Comment by ottok** [comment]
> > I don't understand the commit message / documentation. Why are these changes needed?
> 
> Since MDEV-28153, debian dependencies are now a property of the script. Without adjusting control before `apt-get build-dep`, an incorrect, possibly impossible, set of dependencies are generated.
> 
> > What is broken - what do these changes fix?
> 
> By running the prepare only of autobake-deb.sh, the control file has its build dependencies updated. Running `apt-get build-dep` on any system will now pull the correct acheivable dependencies.

There is no need to run *prepare only*, either you run full autobake-deb.sh or you do a native build. A change based on the above assumption is mislead.

> > What is _early bit of autobake-deb.sh_ that needs lsb-release?
> 
> `CODENAME="$(lsb_release -sc)"`

This command works all the time and is part of standard Debian base. Why add it now?

This PR would be a lot more convincing if it referred a failing CI run on mainline and then a successful one from this PR. That is not the case now.
 
> > What is this SALSA_PREP thing?
> 
> A flag to run the prepare stage only, so that `control` (and `rules`) files only are updated.

Does not make sense as explained above.

> > Why is columnstore stanza refactored?
> 
> It need to be id-potent so that a second run to trigger the build doesn't get two sets of included columnstore files, especially in `debian/control`

This whole section is obsolete now when you changed/broke the previous columnstore packaging structure via ec62f46a612b52aa75d61b971d0a7eb781cc19c1. Either it is part of the native packaging (withfiles in debian/* and stanza in debian/control), or it is not. Any attempt for middle ground just causes a mess and is brings no benefit as far as I can see.

**Comment by grooverdan** [comment]
> > > I don't understand the commit message / documentation. Why are these changes needed?
> > 
> > Since MDEV-28153, debian dependencies are now a property of the script. Without adjusting control before `apt-get build-dep`, an incorrect, possibly impossible, set of dependencies are generated.
> > > What is broken - what do these changes fix?
> > 
> > 
> > By running the prepare only of autobake-deb.sh, the control file has its build dependencies updated. Running `apt-get build-dep` on any system will now pull the correct acheivable dependencies.
> 
> There is no need to run _prepare only_, either you run full autobake-deb.sh or you do a native build. A change based on the above assumption is mislead.
> 
> > > What is _early bit of autobake-deb.sh_ that needs lsb-release?
> > 
> > 
> > `CODENAME="$(lsb_release -sc)"`
> 
> This command works all the time and is part of standard Debian base. Why add it now?
> 
> This PR would be a lot more convincing if it referred a failing CI run on mainline and then a successful one from this PR. That is not the case now.

https://salsa.debian.org/grooverdan/mariadb-server/-/jobs/2603984

> This whole section is obsolete now when you changed/broke the previous columnstore packaging structure via [ec62f46](https://github.com/MariaDB/server/commit/ec62f46a612b52aa75d61b971d0a7eb781cc19c1).

I noticed and corrected this 2885fb0ee496285b49efea7d84c6cd0a8e819319

I'm sorry we have differing opinions on the MDEV-28153 changes.

**Comment by ottok** [comment]
Thanks for https://github.com/MariaDB/server/commit/2885fb0ee496285b49efea7d84c6cd0a8e819319. Please in the future ask for review for all commits before you push them. The PR where you pushed that commit you added those commits after review and proceeded to merge without waiting for second review of the updated version.

I see in https://salsa.debian.org/grooverdan/mariadb-server/-/jobs/2603984  `debian/autobake-deb.sh: line 79: lsb_release: command not found` but this commit is on this PR and I am against this PR, so if this is not merged, the failing build is not introduced in the first place.

To iterate, Salsa-CI builds on branches 10.5/6/7/8/9 are not broken and do not need this fix.

> I'm sorry we have differing opinions on the MDEV-28153 changes.

I still think that the claim about "apt cache being broken" is based on misunderstanding or flawed assumptions. The apt cache is never broken if the builder is healthy and capable of installing new packages. However, since you now refactored that, you should take ownership of the autobake-deb.sh contents. You can do that without having to pollute the salsa-ci.yml file with (at least so far seemingly) unnecessary `SALSA_PREP=1 debian/autobake-deb.sh` lines.

File: debian/autobake-deb.sh
Comment (illuusio) [inline debian/autobake-deb.sh:97]: There could be some comment why this is here as I don't understood reasoning? If we are not in Salsa exit?

Comment (grooverdan) [reply debian/autobake-deb.sh]: `:-` to give it a default value in case its not set. Will prevent failures in bash strict `+x` mode and gets recognised by shellcheck.

**Comment by grooverdan** [comment]
Removed salsa line, Added the one dependency line only.

https://salsa.debian.org/grooverdan/mariadb-server/-/commit/35cb5b446cbbffa6729e5d55d81c224ed56d06d9/pipelines?ref=bb-10.5-danielblack-MDEV-28153-deb-explict-deps-salsa-postfix is the build in progress.

Matched up with https://github.com/MariaDB/mariadb.org-tools/pull/130 build.

File: debian/autobake-deb.sh
Comment (ottok) [reply debian/autobake-deb.sh]: Please add inline comment on what this section is, what uses it and why. Anybody can look up the syntax on a man page but it is hard to guess why you have this seemingly unnecessary logic here.

**Comment by ottok** [comment]
> https://salsa.debian.org/grooverdan/mariadb-server/-/commit/35cb5b446cbbffa6729e5d55d81c224ed56d06d9/pipelines?ref=bb-10.5-danielblack-MDEV-28153-deb-explict-deps-salsa-postfix is the build in progress

Yes, but I am asking where is the build that failed? The addition you did to salsa-ci.yml is unnecessary, it does not do anything. Referencing a build that passes does not prove anything about that that addition fixed (and it didn't fix anything, salsa-ci on 10.5 is already passing).

File: debian/autobake-deb.sh
Comment (ottok) [reply debian/autobake-deb.sh]: Ok, so there should be an inline comment saying that the section is tailored for use with https://github.com/MariaDB/mariadb.org-tools/blob/master/buildbot.mariadb.org/ci_build_images/debian.Dockerfile to make Docker images with preinstalled build-dependencies for each Debian/Ubuntu release, and thus you want to generate the `debian/control` files but not run autobake.

**Comment by illuusio** [comment]
> The addition you did to salsa-ci.yml is unnecessary, it does not do anything.

Why does it not do anything? Is lsb already Salsa-CI dependency? I think point is that it's not breaking anything it's not that it should make it pass?

**Pull request merged**
